### ISSUE TITLE
Merging `help-echo` messages from stacked overlays

### ISSUE THREAD

**Issue by cpitclaudel** [flycheck/flycheck#725]
Hi Sebastian,

At the moment, if a checker reports an info message and a warning for the same buffer location, the warning will take precedence and hide the info message: since the warning overlay has a higher priority, the underline will be that of a warning, and the margin will show a question mark. Unfortunately, that means that the mouseover tooltip (taken from the `help-echo` property of the overlay) will only show the warning message; not the info message. The same thing will happen if two warnings are reported for the same buffer location: only one of them will be shown in the tooltip.

This is exemplified by the following test:

```elisp
(require 'flycheck-ert)

(flycheck-ert-with-temp-buffer
  (insert "a")
  (flycheck-add-overlay (flycheck-error-new-at 1 1 'info "info 1"))
  (flycheck-add-overlay (flycheck-error-new-at 1 1 'info "info 2"))
  (flycheck-add-overlay (flycheck-error-new-at 1 1 'warning "warning 1"))
  (flycheck-add-overlay (flycheck-error-new-at 1 1 'warning "warning 2"))
  (should (string= (help-at-pt-string) "warning 2")))
```

I think that it would nice if the tooptip displayed all error messages instead of just one of the highest priority ones. This behaviour would be consistent with e.g. that of `flycheck-copy-errors-as-kill`. To achieve that, we could modify `flycheck-add-overlay` to set the `help-echo` property to a function instead of an error message. Such a function could be similar to the following:

```elisp
(defun flycheck-help-echo (window object pos)
  "Construct a tooltip message for POS in OBJECT of WINDOW."
  (when (overlayp object)
    (with-selected-window window
      (with-current-buffer (overlay-buffer object)
        (mapconcat #'flycheck-error-message
                   (flycheck-overlay-errors-at pos)
                   "\n\n")))))
```

What do you think? If you like this idea I can prepare a clean PR.

**Comment by cpitclaudel** [comment]
To give credit where due, the idea of using a function for the help-echo comes from [this answer](http://emacs.stackexchange.com/a/14744/2429) :)

**Comment by cpitclaudel** [comment]
Here's a slightly nicer implementation, which preserves the default help-echo values:

```elisp
(defun flycheck-get-help-echo (overlay)
  (or (flycheck-error-message (overlay-get overlay 'flycheck-error))
      (get (overlay-get overlay 'category) 'help-echo)))

(defun flycheck-help-echo (window object pos)
  "Construct a tooltip message for POS in OBJECT of WINDOW."
  (when (overlayp object)
    (with-selected-window window
      (with-current-buffer (overlay-buffer object)
        (-when-let* ((overlays (flycheck-overlays-at pos))
                     (messages (mapcar #'flycheck-get-help-echo overlays)))
          (mapconcat #'identity (delq nil messages) "\n\n"))))))
```

**Comment by lunaryorn** [comment]
@cpitclaudel Wow, now that is what I call a great idea! :+1: :fireworks:   I didn't know that this was possible!  Emacs continues to surprise me every day :smile:

I'd be absolutely happy to merge a PR that implements this.  I think it's going to be a bit more work beyond adapting `flycheck-add-overlay`, because iirc there are a bunch of tests that check the `help-echo` of overlays, but I can do all the grunt work myself if you provide a good implementation for the help-echo function.

A couple of things come to my mind with regards to that implementation:

- I don't think that we still need the default `help-echo` attributes in the overlay categories with this approach, because we can just compute them dynamically, i.e. `(or (flycheck-error-message …) (format "Unknown %s" (flycheck-error-level …))`.  We could summarily remove `help-echo` from the categories.
- I'd be great to make this customizable in a similar way to `flycheck-display-errors-function`.  The `help-echo` function would get all Flycheck errors at point, and then do `(funcall flycheck-help-echo-function errors)`.  The default value of this option would be a function that concatenates all messages like in your example, but users could write their own functions for maximum control over the help echo.

**Comment by cpitclaudel** [comment]
Great! Thanks for the quick answer and for the suggestions. I'll send a PR soon (though it probably won't be before Monday; I'll be pretty busy until then)
